atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,222 @@
+
+    private final Refreshable<Pair<ReadWriteLock, T>> refreshableResourceWithLock;
+    private final Disposable refreshableSubscriptionDisposable;
+
+    @GuardedBy("isClosedLock")
+    private volatile boolean isClosed;
+
+    public static <R, T extends Closeable> ReloadingCloseableContainer<T> of(
+            Function<R, T> resourceFactory, Refreshable<R> refreshableFactoryParameter) {
+        return new ReloadingCloseableContainer<>(resourceFactory, refreshableFactoryParameter);
+    }
+
+    private <R> ReloadingCloseableContainer(
+            Function<R, T> resourceFactory, Refreshable<R> refreshableFactoryParameter) {
+        isClosed = false;
+        refreshableResourceWithLock = refreshableFactoryParameter.map(factoryParameter -> runIfNotClosed(() -> {
+            T resource = resourceFactory.apply(factoryParameter);
+            // This lock does not need to be fair, since the isClosed lock is fair (preventing #close from being
+            // starved) and since subscribe happens after a map (and thus a new resource is created before attempting
+            // to close the old), there will not be any more readers of a resource about to be closed.
+            ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+            return new Pair<>(readWriteLock, resource);
+        }));
+
+        // The old resource is closed in a subscriber (rather than map) so that attempting to close the old resource
+        // happens after the refreshable was updated to use a new resource. This avoids a situation where
+        // a #runWithResource call using a resource that is closed, violating guarantee 1 (between the close lock
+        // being released and map completing the update).
+
+        // Furthermore, we don't want updates to have lock contention with a #runWithResource call:
+
+        // Consider the case of a long-running #runWithResource call. The current resource lock is held.
+        // If we attempt to close the resource in the map, then this must occur before refreshableResourceWithLock is
+        // updated (assuming we don't run close asynchronously). As close holds the write lock for the
+        // resource, subsequent #runWithResource calls would now block on acquiring the lock. This would be doubly
+        // damaging if the resource lock was fair, as now #runWithResource calls are blocked until all calls prior to
+        // closing the resource have finished.
+
+        // To avoid this, we must not block an update with close. This can either be done with another thread, or
+        // simply using a subscriber on the updated resource.
+        refreshableSubscriptionDisposable = refreshableResourceWithLock.subscribe(resourceWithLock -> {
+            Optional<Pair<ReadWriteLock, T>> maybeResourceWithLockToClose =
+                    currentResourceWithLock.getAndSet(Optional.of(resourceWithLock));
+            maybeResourceWithLockToClose.ifPresent(this::closeResourceAndLock);
+        });
```

Comment:
This will not close a resource that is currently serving a request, but as mentioned above, it is possible for this to close a resource that is about to serve a request due to GC. Will this be a problem? Will whatever exception thrown be handled, and/or can you get around this? Hint: I _think_ you can, but it might be enough of an edge case such that retrying based on whatever exception you receive is probably better.